drivers/bluetooth/hci/hci_ifx_cyw208xx.c

```diff
@@ -0,0 +1,392 @@
+
+
+/******************************************************************************
+ *                          Function Declarations
+ ******************************************************************************/
+extern void host_stack_platform_interface_init(void);
+extern void cybt_platform_hci_wait_for_boot_fully_up(bool is_from_isr);
+extern uint8_t *host_stack_get_acl_to_lower_buffer(wiced_bt_transport_t transport, uint32_t size);
+extern wiced_result_t host_stack_send_acl_to_lower(wiced_bt_transport_t transport,
+						   uint8_t *data, uint16_t len);
+extern wiced_result_t host_stack_send_cmd_to_lower(uint8_t *cmd, uint16_t cmd_len);
+extern wiced_result_t host_stack_send_iso_to_lower(uint8_t *data, uint16_t len);
+extern cybt_result_t cybt_platform_msg_to_bt_task(const uint16_t msg, bool is_from_isr);
+extern void cybt_bttask_deinit(void);
+
+static int bt_firmware_download(const uint8_t *firmware_image, uint32_t size)
+{
+	uint8_t *data = (uint8_t *)firmware_image;
+	volatile uint32_t remaining_length = size;
+	struct net_buf *buf;
+	int err;
+
+	LOG_DBG("Executing Fw downloading for CYW208xx device");
+
+	/* The firmware image (.hcd format) contains a collection of hci_write_ram
+	 * command + a block of the image, followed by a hci_write_ram image at the end.
+	 * Parse and send each individual command and wait for the response. This is to
+	 * ensure the integrity of the firmware image sent to the bluetooth chip.
+	 */
+	while (remaining_length) {
+		size_t data_length = data[2]; /* data length from firmware image block */
+		uint16_t op_code = *(uint16_t *)data;
+
+		/* Allocate buffer for hci_write_ram/hci_launch_ram command. */
+		buf = bt_hci_cmd_create(op_code, data_length);
+		if (buf == NULL) {
+			LOG_ERR("Unable to allocate command buffer");
+			return err;
+		}
+
+		/* Add data part of packet */
+		net_buf_add_mem(buf, &data[3], data_length);
+
+		/* Send hci_write_ram command. */
+		err = bt_hci_cmd_send_sync(op_code, buf, NULL);
+		if (err) {
+			return err;
+		}
+
+		switch (op_code) {
+		case BT_HCI_VND_OP_WRITE_RAM:
+			/* Update remaining length and data pointer:
+			 * content of data length + 2 bytes of opcode and 1 byte of data length.
+			 */
+			data += data_length + 3;
+			remaining_length -= data_length + 3;
+			break;
+
+		case BT_HCI_VND_OP_LAUNCH_RAM:
+			remaining_length = 0;
+			break;
+
+		default:
+			return -ENOMEM;
+		}
+	}
+
+	LOG_DBG("Fw downloading complete");
+	return 0;
+}
+
+static int cyw208xx_open(void)
+{
+	int err;
+
+	/* Initialize Bluetooth platform related OS tasks. */
+	cybt_platform_task_init((void *)NULL);
+
+	/* Wait until cybt platform task starts */
+	k_sem_take(&cybt_platform_task_init_sem, K_FOREVER);
+
+	/* Send HCI_RESET */
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_RESET, NULL, NULL);
+	if (err) {
+		return err;
+	}
+
+	/* BT firmware download */
+	err = bt_firmware_download(brcm_patchram_buf, (uint32_t)brcm_patch_ram_length);
+	if (err) {
+		return err;
+	}
```

Comment:
Sorry that I didn't notice this earlier: any HCI command sending needs to happen through a setup() callback. From the host's perspective the HCI transport isn't open until the open() callback has successfully returned, and it's therefore invalid to call APIs like `bt_hci_cmd_send_sync()` before that. It's by chance this works since the host is currently missing checks for this. The setup() driver callback is something that exists specifically for this kind of use cases where you need to do HCI communication (e.g. for firmware download) before the Zephyr Host stack starts sending HCI packets.